collectors/modality-probe-debug-collector/src/lib.rs

```diff
@@ -0,0 +1,1169 @@
+        Self::merge_clock(
+            &mut clocks,
+            LogicalClock {
+                id,
+                epoch: ProbeEpoch(0),
+                ticks: ProbeTicks(0),
+            },
+        );
+        Ok(Self {
+            seq_num: 0,
+            reader: FencedReader::new(
+                MemorySnapper {
+                    mem_accessor,
+                    storage_addr: buf_addr,
+                    write_seqn_high_addr,
+                    write_seqn_low_addr,
+                    overwrite_seqn_high_addr,
+                    overwrite_seqn_low_addr,
+                },
+                buf_cap,
+            ),
+            rbuf: Vec::new(),
+            clocks,
+            priority_writer: PriorityWriter {
+                mem_accessor: priority_mem_accessor,
+                priority_field_addr,
+            },
+        })
+    }
+
+    /// Collect all new logs, return a report
+    pub fn collect_report(&mut self) -> Result<Report, DebugCollectorError> {
+        let num_missed = self.reader.read(&mut self.rbuf)?;
+        // Possibly add entries missed event
+        if num_missed > 0 {
+            let num_missed_rounded = u64::min(num_missed, u32::MAX as u64) as u32;
+            let (ev, payload) =
+                LogEntry::event_with_payload(EventId::EVENT_LOG_ITEMS_MISSED, num_missed_rounded);
+            self.rbuf.insert(0, WholeEntry::Double(ev, payload));
+        }
+
+        // Keep a copy of the clocks before this report which will be used as frontier clocks
+        let report_clocks = self.clocks.clone();
+
+        // Merge clocks from this report
+        for entry in self.rbuf.iter() {
+            if let WholeEntry::Double(first, second) = entry {
+                if first.has_clock_bit_set() {
+                    let id = ProbeId::new(first.interpret_as_logical_clock_probe_id())
+                        .ok_or_else(|| DebugCollectorError::LogProcessingError)?;
+                    let (epoch, ticks) = modality_probe::unpack_clock_word(second.raw());
+                    Self::merge_clock(&mut self.clocks, LogicalClock { id, epoch, ticks })
+                }
+            }
+        }
+
+        let report =
+            Report::try_from_log(self.clocks[0], self.seq_num, report_clocks, &self.rbuf[..])
+                .map(|report| {
+                    self.seq_num += 1;
+                    report
+                })
+                .map_err(|_e| DebugCollectorError::ReportSerializationError);
+        self.rbuf.clear();
+        report
+    }
+
+    fn merge_clock(clocks: &mut Vec<LogicalClock>, ext_clock: LogicalClock) {
+        let mut existed = false;
+        for c in clocks.iter_mut() {
+            if c.id == ext_clock.id {
+                if OrdClock(ext_clock.epoch, ext_clock.ticks) > OrdClock(c.epoch, c.ticks) {
+                    c.epoch = ext_clock.epoch;
+                    c.ticks = ext_clock.ticks;
+                }
+                existed = true;
```

Comment:
I think that we're likely to be maintaining only a single instance of a logical clock per probe id, so it should be fine to `break` after this point.